src/linking/linkModule.ml

```diff
@@ -405,45 +406,50 @@ let set_global global value = fun m ->
   let rec go i = function
     | [] -> assert false
     | g::gs when i = Int32.to_int global ->
-      let open Wasm.Types in
-      assert (g.it.gtype = GlobalType (I32Type, Immutable));
+      let open Wasm_exts.Types in
+      assert (g.it.gtype = GlobalType (I64Type, Immutable));
       let g = phrase (fun g' ->
-        { g' with value = [Const (Wasm.Values.I32 value @@ g.at) @@ g.at] @@ g.at }
+        { g' with value = [Const (Wasm_exts.Values.I64 (Int64.of_int (Int32.to_int value)) @@ g.at) @@ g.at] @@ g.at }
       ) g in
       g :: gs
     | g::gs -> g :: go (i+1) gs
   in
   { m with globals = go 0 m.globals }
 
-let fill_global (global : int32) (value : int32) : module_' -> module_' = fun m ->
-  let rec instr' = function
-    | Block (ty, is) -> Block (ty, instrs is)
-    | Loop (ty, is) -> Loop (ty, instrs is)
-    | If (ty, is1, is2) -> If (ty, instrs is1, instrs is2)
-    | GlobalGet v when v.it = global -> Const (Wasm.Values.I32 value @@ v.at)
+let fill_global (global : int32) (value : Wasm_exts.Values.value) : module_' -> module_' = fun m ->
+  let rec instr' to_32 = function
+    | Block (ty, is) -> Block (ty, instrs to_32 is)
+    | Loop (ty, is) -> Loop (ty, instrs to_32 is)
+    | If (ty, is1, is2) -> If (ty, instrs to_32 is1, instrs to_32 is2)
+    | GlobalGet v when v.it = global -> 
+      let encoded_value = match (to_32, value) with
+      | (true, Wasm_exts.Values.I64 number) -> (Wasm_exts.Values.I32 (Int32.of_int (Int64.to_int number)))
+      | _ -> value in
+      Const (encoded_value @@ v.at)
+    | GlobalGet v when v.it = global -> Const (value @@ v.at)
```

Comment:
This recursive mapping of `to_32` seems odd. What's it doing?

Reply:
This is indeed confusing. I refactored it. 
The reason for `to_32` is to lower the constant expression of table segment offsets to 32-bit. I now kept the original recursive mapping without `to_32` and implemented separate functions for the 32-bit constant expression.